Mathlib/RingTheory/Coalgebra.lean

```diff
@@ -0,0 +1,120 @@
+/-
+Copyright (c) 2023 Ali Ramsey. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ali Ramsey
+-/
+import Mathlib.RingTheory.TensorProduct
+
+/-!
+# Coalgebras
+
+In this file we define `Coalgebra`, and provide instances for:
+
+* Commutative rings: `CommRing.toCoalgebra`
+* Finitely supported functions: `Finsupp.instCoalgebra`
+
+## References
+
+* <https://en.wikipedia.org/wiki/Coalgebra>
+-/
+
+suppress_compilation
+
+universe u v
+
+open scoped TensorProduct
+
+/-- A coalgebra over a commutative ring `R` is a module over `R` equipped with a coassociative
+comultiplication `Δ` and a counit `ε` obeying the left and right conunitality laws. -/
+class Coalgebra (R : Type u) (A : Type v) [CommRing R] [AddCommGroup A] [Module R A] where
+  /-- The comultiplication of the coalgebra -/
+  Δ : A →ₗ[R] A ⊗[R] A
+  /-- The counit of the coalgebra -/
+  ε : A →ₗ[R] R
+  /-- The comultiplication is coassociative -/
+  coassoc : TensorProduct.assoc R A A A ∘ₗ TensorProduct.map Δ .id ∘ₗ Δ =
+    TensorProduct.map .id Δ ∘ₗ Δ
+  /-- The counit satisfies the left counitality law -/
+  ε_id : TensorProduct.lid R A ∘ₗ TensorProduct.map ε .id ∘ₗ Δ = .id
+  /-- The counit satisfies the right counitality law -/
+  id_ε : TensorProduct.rid R A ∘ₗ TensorProduct.map .id ε ∘ₗ Δ = .id
+
+namespace Coalgebra
+
+section CommRingAddCommGroup
+
+variable {R : Type u} {A : Type v}
+variable [CommRing R] [AddCommGroup A] [Module R A] [Coalgebra R A]
```

Comment:
At least for now I think the section `CommRingAddCommGroup` is overkill, I suggest just:
```suggestion
variable {R A : Type*} [CommRing R] [AddCommGroup A] [Module R A] [Coalgebra R A]
```

Reply:
Or even `[CommSemiring R] [AddCommMonoid A]`...

Reply:
I pushed a change with that generalization

Reply:
Thank you!